[Breaking Change] getLedger: include raw transaction with each transaction

intelliot opened the pull request:
getLedger:

* Removes the `rawTransactions` field and replaces it with a new `rawTransaction` field in each transaction.
  * This is consistent with the new `includeRawTransaction` and `includeRawTransactions` options of `getTransaction` and `getTransactions` respectively.
* In each raw transaction, renames the `metaData` field to `meta` for consistency with rippled's `tx` method.
* Also, adds `ledger_index` to each raw transaction.

wilsonianb commented on src/ledger/parse/ledger.ts:
```diff
@@ -25,11 +25,12 @@ export type FormattedLedger = {
 }
```
Remove `rawTransactions` above

intelliot commented:
Looks like `computeLedgerHash` has tests, but when `rawTransactions` is undefined, it just uses the `transactionHash`. Let me try to add a test that forces it to compute a transaction hash instead of just using the one pre-defined... (Is `transactionHash` considered to be part of the "ledger header"?)

intelliot commented:
On `develop`, I added a "hidden" option to `computeLedgerHash` that makes it require raw transactions (it will throw if they are missing). https://github.com/ripple/ripple-lib/commit/3f2d9d198ea5ab9bc24d005ec176239dee018421

On this branch, I updated `computeLedgerHash` to use the new `rawTransaction`s from the new `getLedger` and added a test that it throws when the transaction has been tampered with.

wilsonianb commented:
>Is `transactionHash` considered to be part of the "ledger header"?
yes
https://developers.ripple.com/ledger-header.html#ledger-header

intelliot commented on src/offline/ledgerhash.ts:
```diff
@@ -27,17 +27,29 @@ function hashLedgerHeader(ledgerHeader) {
+  let transactions: any[]
   if (ledger.rawTransactions === undefined) {
     if (options.requireRawTransactions !== true) {
       return ledger.transactionHash
```
Should `headerOnly` default to `true`?

wilsonianb replied:
It looks like previously the presence of `rawState` and `rawTransactions` determined whether we would calculate the corresponding hashes (as opposed to just using the provided hashes to compute the hash of the ledger header).
Despite my earlier suggestion, I'm leaning towards continuing to do the same instead of adding `headerOnly`. The one change would be that we would look for either `rawTransactions` or `transactions`. I think the `ledger is missing raw transactions` ValidationError you added makes sense, since there isn't a way for `getLedger` to return something without either `rawTransactions` or `rawTransaction`s.

intelliot commented on src/offline/ledgerhash.ts:
```diff
@@ -27,17 +27,29 @@ function hashLedgerHeader(ledgerHeader) {
+            ' token u in JSON at position 0') {
+          // one or more of the `tx.rawTransaction`s is undefined
+          throw new common.errors.ValidationError('ledger'
+            + ' is missing raw transactions')
```
Just added in https://github.com/ripple/ripple-lib/pull/935/commits/bb656d38cc282928fc479e230adebdadcb65af7c

intelliot commented:
Rebased and merged in https://github.com/ripple/ripple-lib/commit/76f120bec9dfb25971e2f4dd046cb242ff46b18e
